### DISCUSSION TITLE
Configuration for Batch Run Simulations

### DISCUSSION THREAD

**Discussion by judeurban** [JSBSim-Team/jsbsim#929] [Q&A] [answered]
Hello all!

I've been using JSBSim on a project to study C172 flight trajectories as a result of varying initial conditions. At this stage of the project, batch runs are becoming an increasingly important aspect. However, I'm having trouble properly configuring JSBSim properties between adjacent trials. I've tried a few different approaches so far (which halfway work) but I'd like to do it *correctly*.

# Current Approach

JSBSim has many variables, but two subcategories are "initial variables" and "current" variables. As an example, here are two variables for true airspeed that correspond to the different categories.

```txt
ic/vt-kts                   // represents the initial value of the airspeed, in knots
velocities/vt-fps           // represents the "current" value of the airspeed, in feet per second
```

To move copy the initial conditions into JSBSim, my understanding is that you need something like this.

```cpp
FDM.SetPropertyValue("ic/vt-kts", 90.0);    // sets the initial value of the airspeed to 90 knots
FDM->ResetToInitialConditions(0x0);         // run the initial conditions. This calls FDM->RunIC()
FDM.Run();                                  // run one time step of the flight dynamics model
FDM.GetPropertyValue("velocities/vt-fps");  // returns the "current" value of the airspeed, in feet per second. This should be approximately 90 knots = 151.903 feet per second
```
Assuming that all initial conditions were properly copied, my understanding is that one time step should produce *approximately* the same value. These two returned values should be close. Because of this, all initial conditions variables are essentially buffers to the "current" or "real" variables.

To ensure that these values are being properly copied from the "buffer" initial conditions variables `ic/*` to the "real" variables, I build an "integrity test", which ensures that the difference between `ic/my_buffer_variable` and `propulsion/my_real_variable` are close in value. If the error between these two numbers are above a certain threshold, it indicates that JSBSim has not produced my specified aircraft state, and thus, the batch run is producing invalid data. 

## First Approach

My first approach is to use the `GetPropertyValue()` and `SetPropertyValue()` methods for the `ic/*` variables.

```cpp
FDM.SetPropertyValue("ic/vt-kts", 90.0);    // sets the initial value of the airspeed to 90 knots
FDM->ResetToInitialConditions(0x0);
// or
FDM->RunIC();
```

## Second Approach

The second approach is to obtain the `JSBSim::FGInitialCondition` object from the flight executive, and call specific methods like so.

```cpp
std::shared_ptr<JSBSim::FGInitialCondition> fgic = FDM->GetIC();
fgic->SetVtrueKtsIC(90.0);
FDM->ResetToInitialConditions(0x0);
```

I understand that the `Tie()` method binds the string `"ic/vt-kts"` to the method `SetVtrueKtsIC()`, so these methods do essentially the same thing.

--------------------------

It's also important to keep in mind that I'm wanting to send a complete state, not just true airspeed. I've isolated my variables down to the following.

```python
LATITUDE_DEG
LONGITUDE_DEG
ALTITUDE_FT
KTAS
ROLL_DEG
YAW_DEG
GAMMA
WIND_DIRECTION_DEG
WIND_SPEED_FPS
DENSITY_ALTITUDE_FT
WEIGHT_CONFIGURATION
```

# Current Results

My current results allows for a few (perhaps three) healthy configurations of an aircraft state. After this, my "integrity test" finds massive discrepancies between variables. For example, it finds that true airspeed will be 30 knots lower from its requested value, which can stall the airplane.

## Additional Notes

Without diving incredibly deep into any code, I've identified coupling between different `ic/*` methods. By calling `SetPropertyValue("ic/a", 0.0)`, `GetPropertyValue("ic/b")` will suddenly change. This is obviously a result of the aircraft's cross variable dependencies. 

# Punchline Questions

- How can I provide a set of variables into JSBSim and start the simulation from that configuration?
- Given the list of aircraft state variables above, in which order should I set them so that a cross-dependency doesn't overwrite a value that has already been set?

**Comment by seanmcleod** [comment]
> For example, it finds that true airspeed will be 30 knots lower from its requested value

Are you sure you're not confusing/mixing up TAS and IAS?

**Reply by judeurban** [reply]
Yes I am sure. JSBSim doesn't have indicated airspeed from what I'm seeing. Only true airspeed and calibrated.

I've noticed some cross dependencies in the variables. Setting one influences another. This is obvious when you look at the setters in FGInitialCondition.cpp. Cross variable dependencies. Because of this, the chosen variables (and the order in which you set them) is very important.

I think it would be very useful to document the necessary variables and their correct order. This is something I'm working on now. Hopefully it doesn't exist already!

**Comment by seanmcleod** [comment]
In terms of airspeed you're sure you're not mixing up TAS and CAS (same as IAS just without instrument, position errors)?

I haven't seen it documented explicitly anywhere that I'm aware of, other than in the source code, typically the comment for each method, e.g. 

https://github.com/JSBSim-Team/jsbsim/blob/6c1abbe2ba53268eaf2178bebd946af0e49c7d43/src/initialization/FGInitialCondition.cpp#L748-L752

https://github.com/JSBSim-Team/jsbsim/blob/6c1abbe2ba53268eaf2178bebd946af0e49c7d43/src/initialization/FGInitialCondition.cpp#L620-L624

**Comment by judeurban** [comment]
For this particular batch simulation, here's the psuedo code for what I'm doing. Varying winds out of the north.

- Extracting ALL `ic/*` out of JSBSim (for a default, base reference).
- Loop:
- Set ALL `ic/*` to their base conditions, which were obtained in step 1.
- Setting the "important" variables (the ones I've defined above.) Things like true position, airspeed, winds, attitude, etc.
- Running Flight Dynamics
- Plotting Data
- GOTO: loop

Here's the output. North wind magnitudes vary from 0 to 45 knots. Left Maneuver for a C172. Local easting and northing coordinates.

You can observe

- wind magnitude increasing 
- true airspeed is constant (correct)
- ground speed is decreasing (correct)
- calibrated airspeed is constant (correct)
- trajectory is almost constant (incorrect.) With more winds, I would expect the trajectory to be compressed to the bottom of the map, given that there are winds pushing from the top down.

![top-down-winds](https://github.com/JSBSim-Team/jsbsim/assets/73368004/750e587e-0b5a-45cc-ad33-b64eb5db0088)

There's probably something in `ic/*` that I'm missing.

**Comment by seanmcleod** [comment]
> With more winds, I would expect the trajectory to be compressed to the bottom of the map, given that there are winds pushing from the top down.

But excluding the 45kt wind case, isn't that what your plotted trajectories are showing?

So in the 40kt wind case after the aircraft has travelled 200m to the left it has only travelled ~330m north, but the 0/5kt wind case it has travelled ~390m north.

**Reply by judeurban** [reply]
Interesting. You may be correct. Intuitively, my expectation is that the trajectories would have a larger separation than what's shown. But I could be mistaken.

I'm comparing JSBSim to another flight dynamics model. Perhaps its wind computations are more drastic.

![15-at270](https://github.com/JSBSim-Team/jsbsim/assets/73368004/a6c3da81-1c06-4468-9f24-d7ee372ec2b8)

**Comment by judeurban** [comment, marked as answer]
# Configuring JSBSim for Batch Simulations

Hello all. After much struggle, I am updating this thread because I've discovered a method to do repeatable, reproducible batch simulations in JSBSim. It's actually quite an elegant solution; not too complex.

Here are the results of twenty simulations. Ten were identical (blue), the ten others had random winds (red). You can see IDENTICAL an trajectory in blue, and varied trajectories from random winds.

![new-ic-method](https://github.com/JSBSim-Team/jsbsim/assets/73368004/6b120c67-8e6f-49b1-b27b-8c86a7c40158)

## High level

Here's the basic idea how you can achieve this.

1. Create some initial "base conditions". These are the aircraft state (and environment state) at `t=0` timestep of your simulation. This is what JSBSim will be reset to every time you begin a new trial. For me, I created a JSON file that contains a copy of my initial conditions. I can easily load in all ICs (Initial Conditions) into JSBSim. These variables will eventually populate to their associated `ic/*` counterpart.
3. Initialize JSBSim to these variable values.
4. Declare a new `std::map()` (C++) or `dict()` (Python). Copy ALL `ic/*` variables into the hashmap. Make sure to keep it scoped so that you can use it at the beginning of your simulation. This makes sure that ALL initial conditions are copied from trial to trial.

My python looks something like this.

```python
# get ENTIRE IC state from JSBSim
ic_state_map = jsb.get_initial_conditions()

# ic_state_map looks like this
{
    'ic/alpha-deg': 5.0,
    'ic/alpha-rad': 0.08726646259971647,
    'ic/beta-deg': 0.0,
    'ic/beta-rad': 0.0,
    'ic/gamma-deg': 7.951386703658792e-16,
    'ic/gamma-rad': 1.3877787807814457e-17,
    'ic/h-agl-ft': 4999.975370524856,
    'ic/h-sl-ft': 5000.0,
    'ic/lat-gc-deg': 34.5,
    'ic/lat-gc-rad': 0.6021385919380436,
    'ic/long-gc-deg': -118.00000000000001,
    'ic/long-gc-rad': -2.059488517353309,
    'ic/mach': 0.13845916898601293,
    'ic/p-rad_sec': 0.0,
    'ic/phi-deg': 0.0,
    'ic/phi-rad': 0.0,
    'ic/psi-true-deg': 0.0,
    'ic/psi-true-rad': 0.0,
    'ic/q-rad_sec': 0.0,
    'ic/r-rad_sec': 0.0,
    'ic/roc-fpm': 1.2648451968694188e-13,
    'ic/roc-fps': 2.108075328115698e-15,
    'ic/sea-level-radius-ft': 0.0,
    'ic/terrain-altitude-ft': 0.0,
    'ic/theta-deg': 5.000000000000001,
    'ic/theta-rad': 0.08726646259971649,
    'ic/u-fps': 151.32480004229163,
    'ic/v-fps': 0.0,
    'ic/w-fps': 13.239204514027781,
    'ic/vc-kts': 83.57850748221145,
    'ic/vd-fps': -0.0,
    'ic/ve-fps': 0.0,
    'ic/ve-kts': 83.5450463327801,
    'ic/vg-fps': 151.90283619473269,
    'ic/vg-kts': 89.99996219641588,
    'ic/vn-fps': 151.90283619473269,
    'ic/vt-fps': 151.90283619473269,
    'ic/vt-kts': 89.99996219641588,
    'ic/vw-bx-fps': 0.0,
    'ic/vw-by-fps': 0.0,
    'ic/vw-bz-fps': 0.0,
    'ic/vw-dir-deg': 0.0,
    'ic/vw-down-fps': -2.108075328115698e-15,
    'ic/vw-east-fps': 0.0,
    'ic/vw-mag-fps': 5.684341886080802e-14,
    'ic/vw-north-fps': 5.684341886080802e-14
}
```

Now, you can do this to reset JSBSim for a batch trial.

```python
for i in range(num_trials):

    # reset all initial conditions to how they were at `t=0`
    jsb.set_initial_conditions(ic_state_map)

    # modify any parameters you want here (optional)
    jsb.set_property_value("ic/phi-deg", 0)
    jsb.set_property_value("ic/theta-deg", -5)
    jsb.set_property_value("ic/psi-true-deg", 0)
    jsb.set_property_value("ic/u-fps", 150)
    jsb.set_property_value("ic/v-fps", 0)
    jsb.set_property_value("ic/w-fps", 0)

    # set the simulation to the initial conditions that you have just specified
    # This method calls FDM->ResetToInitialConditions(0x0), which initializes all models from scratch and calls RunIC()
    jsb.reset_to_initial_conditions()
```

I should also mention that I did not use JSBSim Python for any of this. Long story, so I won't go into details. That being said, they methods names may not match exactly. I'm just presenting the basic idea for everyone here.
